staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/list_width.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"math"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+const (
+	// the minimum number of seats a list request occupies
+	minimumSeatsList uint = 3
+
+	// the maximum number of seats a list request occupies
+	maximumSeatsList uint = 10
+)
+
+func newListWidthEstimator(countFn objectCountGetterFunc) WidthEstimatorFunc {
+	estimator := &listWidthEstimator{
+		countFn: countFn,
+	}
+	return estimator.estimate
+}
+
+type listWidthEstimator struct {
+	countFn objectCountGetterFunc
+}
+
+func (e *listWidthEstimator) estimate(r *http.Request) Width {
+	requestInfo, ok := apirequest.RequestInfoFrom(r.Context())
+	if !ok {
+		return Width{Seats: minimumSeatsList}
+	}
+
+	query := r.URL.Query()
+	count := e.countFn.Get(key(requestInfo))
+
+	var fieldSelectorPresent bool
+	if len(query.Get("fieldSelector")) > 0 {
+		fieldSelectorPresent = true
+	}
+
+	var resourceVersionNotSet bool
+	if version, ok := query["resourceVersion"]; ok && len(version) > 0 && len(version[0]) == 0 {
+		resourceVersionNotSet = true
+	}
+
+	limit := parseLimit(query)
+
+	var estimatedObjectsToBeProcessed int64
+	switch {
```

Comment:
In the initial version what we should do imho is to do:
```
switch {
case isListFromCache:
  estimatedObjectsToBeProcessed = count
case hasSelector:
  estimatedObjectsToBeProcessed = count
default:
  estimatedObjectsToBeProcesed = limit
}
```

Reply:
Now, actually thinking more about this...

If the selector is actually set, we indeed may need to process many more objects. But all chunks we process are of size <limit> and the actual total number of objects will also be of size limit.

So thinking more about this - it shouldn't actually influcence the width - it will be somehow automatically counted in the processing latency of the request.


So I think it should actually go something like this:
```
switch {
case isListFromCache:
  estimatedObjectsToBeProcessed = count
default:
  // Even if a selector is specified and we may need to list and go over more objects from etcd
  // to produce the result of size <limit>, each individual chunk will be of size at most <limit>.
  // As a result. the width of the request should be computed based on <limit> and the actual
  // cost of processing more elements will be hidden in the request processing latency.
  estimatedObjectsToBeProcessed = limit
}
```

Reply:
And add a
```
// TODO: For resources that implement indexes at the watchcache level, we need to adjust the cost accordingly
```

Reply:
I actaully did part of that in https://github.com/kubernetes/kubernetes/pull/103539

Feel free to copy from there.